File: GSP_test/GSPtest/LogReader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace GSPtest
{
  class Action
  {
    public Action(int _time, string _command)
    {
      this.time = _time;
      this.command = _command;
    }
    public int time { get; set; }
    public string command { get; set; }
  }
  class LogReader
  {
    public List<string> commands = new List<string>();
    public List<List<Action>> sessions = new List<List<Action>>();
    public void ReadLog(string file_name)
    {
      using (StreamReader stream = new StreamReader(file_name))
      {
        List<Action> log = new List<Action>();
        string line = "";
        while ((line = stream.ReadLine()) != null)
        {
          int time = GetTimeFromRecord(line);
          if (time > 0)
          {
            if (GetCommandFromRecord(line, out string cmd))
            {
              bool cmd_exists = false;
              foreach (string c in commands)
                if (c == cmd)
                {
                  cmd_exists = true;
                  break;
                }
              if (!cmd_exists)
                commands.Add(cmd);
              log.Add(new Action(time, cmd));
            }
          }
        }
        sessions.Add(log);
      }
    }
    public void ReadLog2(string file_name)
    {
      using (StreamReader stream = new StreamReader(file_name))
      {
        List<Action> log = new List<Action>();
        string line = "";
        while ((line = stream.ReadLine()) != null)
        {
          int time = GetTimeFromRecord(line);
          if (time > 0)
          {
            if (GetCommandFromRecord(line, out string cmd))
            {
              bool cmd_exists = false;
              foreach (string c in commands)
                if (c == cmd)
                {
                  cmd_exists = true;
                  break;
                }
              if (!cmd_exists)
                commands.Add(cmd);
              log.Add(new Action(time, cmd));
              if (cmd == "Exit")
              {
                sessions.Add(log);
                log = new List<Action>();
              }
            }
          }
        }
        sessions.Add(log);
      }
    }
    public void ReadDir(string path)
    {
      string[] files = Directory.GetFiles(path, "*.log");
      foreach (string file in files)
        this.ReadLog2(file);
    }
    public void GetDataForGPS(out List<Client> clients, out Dictionary<int, string> item_map)
    {
      clients = new List<Client>();
      item_map = new Dictionary<int, string>();
      int id = 1;
      foreach (string cmd in commands)
        item_map.Add(id++, cmd);
      for (int i = 0; i < sessions.Count; i++)
      {
        List<Transaction> trans = new List<Transaction>();
        foreach (Action act in sessions[i])
          trans.Add(new Transaction(i, act.time, new ItemGroup(item_map.FirstOrDefault(x => x.Value == act.command).Key)));
        Client client = new Client(i);
        client.SetTransactions(trans, item_map.Count);
        clients.Add(client);
      }
    }
    private int GetTimeFromRecord(string r)
    {
      if (r.Length < 20)
        return -1;
      if (!Int32.TryParse(r.Substring(0, 4), out int year))
        return -1;
      if (!Int32.TryParse(r.Substring(5, 2), out int month))
        return -1;
      if (!Int32.TryParse(r.Substring(8, 2), out int day))
        return -1;
      if (!Int32.TryParse(r.Substring(11, 2), out int hours))
        return -1;
      if (!Int32.TryParse(r.Substring(14, 2), out int minutes))
        return -1;
      if (!Int32.TryParse(r.Substring(17, 2), out int seconds))
        return -1;

      int res = seconds + minutes * 60 + hours * 60 * 60 + day * 24 * 60 * 60;
      return res;
    }
    private bool GetCommandFromRecord(string r, out string res)
    {
      res = "";
      int start = r.IndexOf('<')+1;
      int end = r.IndexOf('>');
      if (start == 0 || end == 0)
        return false;
      res = r.Substring(start, end - start);
      return true;
    }
  }
}

File: GSP_adapt/LogReader.cpp
#include "LogReader.h"
#include "DataBase.h"

LogReader::LogReader()
{
  dir_filters = QDir::Files | QDir::Dirs | QDir::NoDotAndDotDot;
  ignore_commands = {"Inspector", "StartupVPerfTest", "TipOfDay",
                     "/Inspector", "/StartupVPerfTest", "/TipOfDay",
                     "/Exit", "/NewDocument", "/OpenDocument"};
  new_session_commands = {"Exit", "NewDocument", "OpenDocument"};
  end_cmds = false;
}

shared_ptr<LogReader> LogReader::instance()
{
  static shared_ptr<LogReader> my_instance(new LogReader());

  return my_instance;
}

void LogReader::readLogs(QString dir_name)
{
  QDir dir(dir_name);
  dir.setFilter(dir_filters);
  //dir.setNameFilters(name_filters);
  QList<QString> commands;
  int session_id = 0;

  for (QFileInfo& file_info : dir.entryInfoList())
    if (file_info.isFile() && file_info.completeSuffix() == "log")
      readFile(file_info, commands, session_id);
    else if (file_info.isDir())
      readDir(file_info.absoluteFilePath(), commands, session_id);
}

void LogReader::readLogsWithoutTime(QString dir_name)
{
  QDir dir(dir_name);
  dir.setFilter(dir_filters);
  //dir.setNameFilters(name_filters);
  QList<QString> commands;
  int session_id = 0;

  for (QFileInfo& file_info : dir.entryInfoList())
  {
    if (file_info.isFile() && file_info.completeSuffix() == "log")
    {
      readFileWithoutTime(file_info, commands, session_id);
    }
    else if (file_info.isDir())
    {
      readDirWithoutTime(file_info.absoluteFilePath(), commands, session_id);
    }
  }
}

void LogReader::includeEndCmds(bool val)
{
  LogReader::end_cmds = val;
}

QStringList LogReader::getIgnoreList() const
{
  return ignore_commands;
}

QStringList LogReader::getNewSessionCmdsList() const
{
  return new_session_commands;
}

void LogReader::setIgnoreList(QStringList list)
{
  ignore_commands = list;
}

void LogReader::setNewSessionCmdsList(QStringList list)
{
  new_session_commands = list;
}

void LogReader::readFile(const QFileInfo &file_info, QList<QString> &commands, int &session_id)
{
  QFile file (file_info.absoluteFilePath());
  //qDebug() << "File:" << "\t" << file.fileName();

  if (!file.open(QIODevice::ReadOnly))
    return; // не получилось открыть файл

  shared_ptr<DataBase> db = DataBase::instance();
  bool empty_session = true;
  while (!file.atEnd())
  {
    QString line = QString(file.readLine());
    int time = getTimeFromRecord(line);
    if (time > 0)
    {
      QString datetime = line.left(19);
      if (end_cmds)
      {
        QStringList cmds = getTwoCommandsFromRecord(line);
        if (cmds.size() > 2)
          qDebug () << "Here: " << cmds;
        for (int i = 0; i < cmds.size(); i++)
        {
          QString cmd = cmds[i];
          if (new_session_commands.contains(cmd))
          {
            if (!empty_session)
            {
              session_id++;
              empty_session = true;
            }
          }
          else
          {
            empty_session = false;
            bool cmd_exists = false;
            for (const QString& c : commands)
              if (c == cmd)
              {
                cmd_exists = true;
                break;
              }
            if (!cmd_exists)
              commands.append(cmd);

            int id = -1;
            if (db->addCommand(session_id, datetime, cmd, id) != OK)
              qDebug() << db->lastError(); //Make throw instead qDebug()?
            if (id % 100 == 0)
              qDebug() << id;
          }
        }
      }
      else
      {
        QString cmd = "";
        if (getCommandFromRecord(line, cmd))
        {
          if (new_session_commands.contains(cmd))
          {
            if (!empty_session)
            {
              session_id++;
              empty_session = true;
            }
          }
          else
          {
            empty_session = false;
            bool cmd_exists = false;
            for (const QString& c : commands)
              if (c == cmd)
              {
                cmd_exists = true;
                break;
              }
            if (!cmd_exists)
              commands.append(cmd);

            int id = -1;
            if (db->addCommand(session_id, datetime, cmd, id) != OK)
              qDebug() << db->lastError(); //Make throw instead qDebug()?
            if (id % 1000 == 0)
              qDebug() << id;
          }
        }
      }
    }
  }

  file.close();
}

void LogReader::readDir(const QString& abs_path, QList<QString> &commands, int &session_id)
{
  QDir dir(abs_path);
  dir.setFilter(dir_filters);
  //dir.setNameFilters(name_filters);

  for (QFileInfo& file_info : dir.entryInfoList())
  {
    if (file_info.isFile() && file_info.completeSuffix() == "log")
    {
      readFile(file_info, commands, session_id);
    }
    else if (file_info.isDir())
    {
      readDir(file_info.absoluteFilePath(), commands, session_id);
    }
  }
}

void LogReader::readFileWithoutTime(const QFileInfo &file_info, QList<QString> &commands, int &session_id)
{
  QFile file (file_info.absoluteFilePath());
  //qDebug() << "File:" << "\t" << file.fileName();

  if (!file.open(QIODevice::ReadOnly))
    return; // не получилось открыть файл

  shared_ptr<DataBase> db = DataBase::instance();
  int int_time = 0;
  while (!file.atEnd())
  {
    QString line = QString(file.readLine());
    if (line[0] == '<')
    {
      QString cmd;
      if (getCommandFromRecord(line, cmd))
      {
        bool cmd_exists = false;
        for (const QString& c : commands)
          if (c == cmd)
          {
            cmd_exists = true;
            break;
          }
        if (!cmd_exists)
          commands.append(cmd);

        if (db->addCommand(session_id, int_time++, cmd) != OK)
          qDebug() << db->lastError(); //Make throw instead qDebug()?

        if (cmd == "Exit")
        {
          session_id++;
          int_time = 0;
        }
      }
    }
  }

  file.close();
}

void LogReader::readDirWithoutTime(const QString &abs_path, QList<QString> &commands, int &session_id)
{
  QDir dir(abs_path);
  dir.setFilter(dir_filters);
  //dir.setNameFilters(name_filters);

  for (QFileInfo& file_info : dir.entryInfoList())
  {
    if (file_info.isFile() && file_info.completeSuffix() == "log")
    {
      readFileWithoutTime(file_info, commands, session_id);
    }
    else if (file_info.isDir())
    {
      readDirWithoutTime(file_info.absoluteFilePath(), commands, session_id);
    }
  }
}

int LogReader::getTimeFromRecord(QString r)
{
  if (r.length() < 20)
    return -1;
  bool ok = false;
  int year = r.mid(0, 4).toInt(&ok);
  if (!ok)
    return -1;
  int month = r.mid(5, 2).toInt(&ok);
  if (!ok)
    return -1;
  int day = r.mid(8, 2).toInt(&ok);
  if (!ok)
    return -1;
  int hours = r.mid(11, 2).toInt(&ok);
  if (!ok)
    return -1;
  int minutes = r.mid(14, 2).toInt(&ok);
  if (!ok)
    return -1;
  int seconds = r.mid(17, 2).toInt(&ok);
  if (!ok)
    return -1;

  int res = seconds + minutes * 60 + hours * 60 * 60 + day * 24 * 60 * 60;
  return res;
}

bool LogReader::getCommandFromRecord(QString r, QString &res)
{
  res = "";
  int start = r.indexOf('<')+1;
  int end = r.indexOf('>');
  if (start == 0 || end == 0)
    return false;
  res = r.mid(start, end - start);

  if (ignore_commands.contains(res))
    return false;

  return true;
}

//Считывает <> также внутри стркои команды
QStringList LogReader::getAllCommandsFromRecord(QString r)
{
  QStringList res;
  int cmds_n = r.count("<");

  int start = -1;
  int end = -1;
  for (int i = 0; i < cmds_n; i++)
  {
    QString s = "";
    start = r.indexOf('<', start + 1)+1;
    if (r[start] == '/')
      end = r.indexOf(':', start + 1);
    else
      end = r.indexOf('>', start + 1);
    if (start == -1 || end == -1)
      continue;
    s = r.mid(start, end - start);
    if (ignore_commands.contains(s))
      continue;
    res.append(s);
  }
  return res;
}

//Считывает только начало <> и конец </> команд
QStringList LogReader::getTwoCommandsFromRecord(QString r)
{
  QStringList res;
  QString s = "";
  int start = r.indexOf('<');
  int end = r.indexOf('>');
  if (start != -1 && end != -1)
  {
    start++;
    s = r.mid(start, end - start);
    if (!ignore_commands.contains(s))
      res.append(s);
  }

  s = "";
  start = r.indexOf("</", start + 1);
  end = r.indexOf(':', start + 1);
  if (start != -1 && end != -1)
  {
    start++;
    s = r.mid(start, end - start);
    if (!ignore_commands.contains(s))
      res.append(s);
  }

  return res;
}
